pkg/datapath/linux/bigtcp/bigtcp.go

```diff
@@ -0,0 +1,152 @@
+	return c.gsoMaxSize
+}
+
+func setGROGSOMaxSize(device string, GROMaxSize, GSOMaxSize int) error {
+	link, err := netlink.LinkByName(device)
+	if err != nil {
+		log.WithError(err).WithField("device", device).Warn("Link does not exist")
+		return nil
+	}
+
+	attrs := link.Attrs()
+	if (int(attrs.GROMaxSize) == GROMaxSize && int(attrs.GSOMaxSize) == GSOMaxSize) ||
+		(!option.Config.EnableIPv6BIGTCP &&
+			int(attrs.GROMaxSize) <= GROMaxSize &&
+			int(attrs.GSOMaxSize) <= GSOMaxSize) {
+		return nil
+	}
+
+	err = netlink.LinkSetGROMaxSize(link, GROMaxSize)
+	if err != nil {
+		return err
+	}
+
+	return netlink.LinkSetGSOMaxSize(link, GSOMaxSize)
+}
+
+func InitBIGTCP(bigtcpConfig *Configuration) {
+	var err error
+
+	if option.Config.DryMode {
+		return
+	}
+
+	if len(option.Config.GetDevices()) == 0 {
+		if option.Config.EnableIPv6BIGTCP {
+			log.Warn("IPv6 BIG TCP could not detect host devices. Disabling the feature.")
+		}
+		option.Config.EnableIPv6BIGTCP = false
+		return
+	}
+
+	kernelGood := false
+	if h := probes.NewProbeManager().GetHelpers("sched_cls"); h != nil {
+		_, kernelGood = h["bpf_dynptr_data"]
+	}
+	if !kernelGood {
+		if option.Config.EnableIPv6BIGTCP {
+			log.Fatalf("Cannot enable --%s, needs kernel 5.19 or newer",
+				option.EnableIPv6BIGTCP)
+		}
+		return
+	}
+
+	disableMsg := ""
+	bigtcpConfig.groMaxSize = defaultGROMaxSize
+	bigtcpConfig.gsoMaxSize = defaultGSOMaxSize
+	if option.Config.EnableIPv6BIGTCP {
+		if option.Config.DatapathMode != datapathOption.DatapathModeVeth {
+			log.Fatal("IPv6 BIG TCP is supported only in veth datapath mode")
+		}
+		if option.Config.TunnelingEnabled() {
+			log.Fatal("IPv6 BIG TCP is not supported in tunneling mode")
+		}
+		if option.Config.EncryptionEnabled() {
+			log.Fatal("IPv6 BIG TCP is not supported with encryption enabled")
+		}
+		if option.Config.EnableHostLegacyRouting {
+			log.Fatal("IPv6 BIG TCP is not supported with legacy host routing")
+		}
+
+		log.Info("Setting up IPv6 BIG TCP")
+		bigtcpConfig.groMaxSize = BIGTCPGROMaxSize
+		bigtcpConfig.gsoMaxSize = BIGTCPGSOMaxSize
+		disableMsg = ", disabling BIG TCP"
+	}
+
+	modifiedDevices := []string{}
+	for _, device := range option.Config.GetDevices() {
+		// we always add the device because we might do only a partial
+		// modification and end up with an error, so best to be conservative
+		// and always reset all on error
+		modifiedDevices = append(modifiedDevices, device)
+		err = setGROGSOMaxSize(device, bigtcpConfig.groMaxSize, bigtcpConfig.gsoMaxSize)
+		if err != nil {
+			log.WithError(err).WithField("device", device).Warnf("Could not modify gro_max_size and gso_max_size%s", disableMsg)
+			option.Config.EnableIPv6BIGTCP = false
+			break
+		}
+		log.WithField("device", device).Infof("Setting gso_max_size to %d and gro_max_size to %d",
+			bigtcpConfig.gsoMaxSize, bigtcpConfig.groMaxSize)
+	}
+
+	if err != nil {
+		bigtcpConfig.groMaxSize = defaultGROMaxSize
+		bigtcpConfig.gsoMaxSize = defaultGSOMaxSize
```

Comment:
Looks like default*Size is just a var  const. Probably to be save should read these out at top of function so if something else has configured values we don't whack them here on error. GKE for example does some /proc settings I would guess others do as well.

Reply:
the problem is once it's enabled we don't know what it was before that, we could record them but it seems like an overkill, gso and gro max sizes are rarely changed from their 64k defaults,
I'll think about it more and will check if any of the vendors changes them

Reply:
By the way note that these settings are per-device and can only be changed via netlink calls (they aren't exported to /proc or /sys)

Reply:
I left a comment in big tcp's docs about the assumptions we make and the logic of enable/disable (when Cilium actually does any value changes). I think that should be ok for the time being, if we hit any issues we can revisit the logic and perhaps persist the old values to be sure that we need to change them to anything different.